### ISSUE TITLE
Not getting 120fps on Samsung Galaxy S20+

### ISSUE THREAD

**Issue by Nogway** [ppy/osu#10050]
**Describe the bug:** 
I can't get 120fps on my Samsung Galaxy S20+, I tried to mess arround with in game video settings but it doesn't matter, still stuck arround 60fps. Itr tried Vsync, 2x refresh rate, 4x refresh rate, 8x and unlimited but I'm sure that my phone get stuck at 60Hz, I can see it when I try to scroll the top bar. I tried to find something on the internet to force my phone to stay at 120Hz but people say that the game is the only problem...

Android version : 10 (G986BXXU4BTH5)

**Screenshots or videos showing encountered issue:** 
Ask me any you need, I will answer with screenshots.

**osu!lazer version:** 
Osu!lazer version : 2020.903.0

**Logs:**
Not needed

**Comment by bdach** [comment]
As a sanity check please see if [this guide](https://www.androidcentral.com/how-change-galaxy-s20-refresh-rate-120hz-or-60hz) does anything for you. I'm not aware of anything that would be limiting to 60fps on Android explicitly on our side.

**Comment by Nogway** [comment]
> As a sanity check please see if [this guide](https://www.androidcentral.com/how-change-galaxy-s20-refresh-rate-120hz-or-60hz) does anything for you. I'm not aware of anything that would be limiting to 60fps on Android explicitly on our side.

Hello, thank you for the link. These steps are already done on my phone, for example I can play subway surfer on my phone with 120hz. Here is a link containing every games working on 120hz phones : [www.androidpolice.com](https://www.androidpolice.com/2020/02/15/these-high-fps-games-are-ready-for-the-galaxy-s20s-120hz-screen/)

**Comment by bdach** [comment]
Thanks for confirming. As this issue requires owning android hardware that can support 120fps to debug, I unfortunately can't guarantee that this will be supported anytime soon.

**Comment by smoogipoo** [comment]
It's probably going to improve once we fully switch to SDL.

**Comment by Susko3** [comment]
I'm having the same issue with Realme 6 (90 Hz screen): osu! runs with 60 fps no matter what settings I try.

I've noticed something weird. When I begin opening the app switcher (swipe from the bottom of the screen and hold), osu! will run with 90 FPS (as seen on the fps meter). So osu! can run in 90 Hz on Android. The 90 Hz only works when I'm in the app switcher animation, as soon as I lift my finger off the screen, the framerate returns to 60 fps.

### What I think is happening:

- Something in the way osu-framework/osuTK draws on Android results in requesting a 60 Hz refresh rate, so Android sets the screen refresh rate to 60 Hz
- As soon as another more important app (the Android app  switcher) requests high refresh rate, Android will use 90 Hz mode, and osu! adopts to the new framerate and displays in 90 fps
  - the app switcher requests/forces 90 Hz so the animation can be smooth

(Hope my explanation makes sense, I can post a screen recording if necessary.)


I've got some data to support my thinking. From `adb shell dumpsys display` command:

Just osu! running normally:
```
DisplayModeDirector
  mSupportedModesByDisplay:
    0 -> [{id=1, width=1080, height=2400, fps=90.0}, {id=2, width=1080, height=2400, fps=60.0}]
  mDefaultModeByDisplay:
    0 -> {id=2, width=1080, height=2400, fps=60.0}
  mVotesByDisplay:
    -1:
      PRIORITY_USER_SETTING -> Vote{width=-1, height=-1, minRefreshRate=0.0, maxRefreshRate=60.0}
    0:
      PRIORITY_APP_REQUEST_SIZE -> Vote{width=1080, height=2400, minRefreshRate=0.0, maxRefreshRate=Infinity}
      PRIORITY_APP_REQUEST_REFRESH_RATE -> Vote{width=-1, height=-1, minRefreshRate=60.0, maxRefreshRate=60.0}
  SettingsObserver
    mDefaultPeakRefreshRate: 60.0
  AppRequestObserver
    mAppRequestedModeByDisplay:
    0 -> {id=2, width=1080, height=2400, fps=60.0}
```

With app switcher open:
```
DisplayModeDirector
  mSupportedModesByDisplay:
    0 -> [{id=1, width=1080, height=2400, fps=90.0}, {id=2, width=1080, height=2400, fps=60.0}]
  mDefaultModeByDisplay:
    0 -> {id=2, width=1080, height=2400, fps=60.0}
  mVotesByDisplay:
    -1:
      PRIORITY_USER_SETTING -> Vote{width=-1, height=-1, minRefreshRate=0.0, maxRefreshRate=60.0}
    0:
      PRIORITY_APP_REQUEST_SIZE -> Vote{width=1080, height=2400, minRefreshRate=0.0, maxRefreshRate=Infinity}
      PRIORITY_APP_REQUEST_REFRESH_RATE -> Vote{width=-1, height=-1, minRefreshRate=90.0, maxRefreshRate=90.0}
  SettingsObserver
    mDefaultPeakRefreshRate: 60.0
  AppRequestObserver
    mAppRequestedModeByDisplay:
    0 -> {id=1, width=1080, height=2400, fps=90.0}
```

Note the line `PRIORITY_APP_REQUEST_REFRESH_RATE `

@Nogway can you try and reproduce 120 fps on your phone, as I described above.

**Comment by Nogway** [comment]
@Susko3 I confirm that I can get 120 fps as you did.

Stuck at 60fps in fullscreen :
![Screenshot_20201124-141134](https://user-images.githubusercontent.com/70708402/100099541-df6a7980-2e5f-11eb-80be-9612bc832562.jpg)

Can get up to 120fps while doing the trick :
![Screenshot_20201124-141159](https://user-images.githubusercontent.com/70708402/100099683-16408f80-2e60-11eb-9e18-2fa1c726b83b.jpg)

**Comment by Susko3** [comment]
> It's probably going to improve once we fully switch to SDL.

Unfortunately, even after the SDL2 update (2020.1203), 90 fps still not working when the app is in focus. (Can still get 90 fps in app switcher animation)

**Comment by peppy** [comment]
we are not using SDL2 for mobile yet, so nothing should have changed.

**Comment by 9hax** [comment]
I can reproduce the App Switcher thing on my OnePlus Nord.

**Comment by Susko3** [comment]
My phone (Realme 6) was just upgraded to Android 11, and setting the refresh rate in system Settings to 90 Hz actually uses that refresh rate in osu!(framework). So this issue was fixed for me in a software update.

Something to note: 
In Settings, I have three options: "90 Hz", "Auto-select", and "60 Hz"
If I select "Auto-select" as the refresh rate, osu! will use 60 Hz.

**Comment by Nogway** [comment]
Well, I've enabled the dev mode on my phone and enable the "Show refresh frequency" thing and it looks like the problem showing the game at with 120HZ is no more (it just looks like the performance are just not good enough, maybe the gpu just can't handle the work). I've got same results for using "vsync", "2x refresh rate", "4x refresh rate", "8x refresh rate" and "unlimited". But the performances are a bit worse while using vsync than any other available. 

Here are the screenshots (the green "120" indicator is from the settings I checked in the dev options and NOT an app showing the refresh rate of my screen) : 

![Screenshot_20210912-030133](https://user-images.githubusercontent.com/70708402/132967438-92a76362-0c67-4d84-a0cc-07d1e0390d27.jpg)

![Screenshot_20210912-030005](https://user-images.githubusercontent.com/70708402/132967443-9f99d508-1dab-432d-8592-956877b94b6d.jpg)

As you can see as well, I tried rotating the screen and it looks like it doesn't affect the performance at all. I tried to check if forcing it in fullscreen would change anything, the result is that it doesn't. Tried with the "Fullscreen apps" in my phone settings.

Using the latest version available as today : 2021.907.0-lazer 

Closed case or is the performance issue linked to that thing ?
